Fog effect when player is inside cloud

bendeutsch opened the pull request:
This pull request detects when a player (more specifically, the camera) is inside a cloud and adjusts the fog, sky and background color accordingly.

This is in response to issue #3576 .

Note that this is currently a work in progress; as such, it may be a bit early to discuss code style.

Some notes:
- Works in 1st person and 3rd person camera mode.
- This is a 100% client-side effect; no server communication or specific version required.
- Since it uses fog, the effect is completely negated when fog is turned off.
- There is also no effect if 3D clouds are disabled, as you can't be "in a cloud" then.

bendeutsch commented:
At the moment, there is a working effect when shaders are disabled. I concentrated on the branch without shaders first, since the fog parameters are more readily accessible there. I got the idea of adjusting sky and background colors only recently, so it may in fact work with shaders enabled already, but I haven't even tried it so far.

nerzhul commented on src/clouds.cpp:
```diff
@@ -23,13 +23,17 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 irr::scene::ISceneManager *g_menucloudsmgr = NULL;
 
+// Constant for now
+static const float cloud_size = BS * 64.0f;
```
static constexpr const

bendeutsch replied:
Done

nerzhul commented on src/clouds.cpp:
```diff
@@ -44,7 +48,12 @@ Clouds::Clouds(
 	scene::ISceneNode(parent, mgr, id),
-	m_seed(seed)
+	m_seed(seed),
+	m_camera_pos(0.0f, 0.0f, 0.0f),
```
please use proper C++11 constructors instead of old constructors for hardcoded values (see recent PR #5999)

bendeutsch replied:
Done

nerzhul commented on src/clouds.cpp:
```diff
@@ -113,19 +123,19 @@ void Clouds::render()
 	v2s16 center_of_drawing_in_noise_i(
-		MYROUND(center_of_drawing_in_noise_f.X / cloud_size),
-		MYROUND(center_of_drawing_in_noise_f.Y / cloud_size)
+		floor(center_of_drawing_in_noise_f.X / cloud_size),
```
std::floor

bendeutsch replied:
`error: ‘floor’ is not a member of ‘std’` – I had included `<math.h>`, it works with `<cmath>`. Hope that's ok?

bendeutsch replied:
Using `<cmath>` now.

nerzhul commented on src/clouds.cpp:
```diff
@@ -361,6 +361,22 @@ void Clouds::update(v2f camera_p, video::SColorf color_diffuse)
+	// is the camera inside the cloud mesh?
+	m_camera_inside_cloud = false; // default
+	if (m_enable_3d)
+	{
```
brace style is wrong everywhere

bendeutsch replied:
Yes, sorry, I did mention it was still in the early churning phase. I use vim, so getting clang-format for only a few lines is often tricky, which is why I save it for later.

bendeutsch replied:
Ran clang_format over all major changed lines.

nerzhul commented on src/clouds.cpp:
```diff
@@ -351,7 +351,7 @@ void Clouds::step(float dtime)
 }
 
-void Clouds::update(v2f camera_p, video::SColorf color_diffuse)
+void Clouds::update(v3f camera_p, video::SColorf color_diffuse)
```
v3f should be passed as const ref, video::SColorf too

bendeutsch replied:
Done

nerzhul commented on src/clouds.cpp:
```diff
@@ -370,3 +386,16 @@ void Clouds::readSettings()
+			(float)y * cloud_size_noise,
+			m_seed, 3, 0.5);
+	// normalize to 0..1 (given 3 octaves)
+	static const float noise_bound = 1.0f + 0.5f + 0.25f;
```
static constexpr const

bendeutsch replied:
Done.

nerzhul commented on src/clouds.h:
```diff
@@ -118,6 +118,16 @@ class Clouds : public scene::ISceneNode
 		updateBox();
 	}
 
+	bool isCameraInsideCloud()
```
getters are const functions

bendeutsch replied:
Done

bigfoot547 commented:
Do you have screenshots? Or does this only detect if a player is in a cloud and not add any visual effects yet?

paramat commented:
Why is it necessary to change sky and background colours? It seems that a short-range normal fog is fine as the only effect.

bendeutsch commented:
@bigfoot547 : no, this currently works. I'm tweaking the fog parameters at the moment, and also refactoring the way the effect is achieved. Screenshots should follow soon.

@paramat : recall that the "sky" is just basically the clear color of the color buffer. In OpenGL and similar libraries, fog affects only elements that are actually drawn. This would have all nodes and entities quickly fade to white – against a bright blue sky. I'll see if I can mock up a screenshot of that, too.

paramat commented:
I see.

bendeutsch commented:
**Screenshots**:

Inside a cloud (shaders enabled):
![clouds_inside](https://user-images.githubusercontent.com/2880643/27316950-8f4fcc60-5583-11e7-8439-998acfaa3d63.png)

Next to cloud for comparison:
![clouds_outside](https://user-images.githubusercontent.com/2880643/27316921-6d757cfc-5583-11e7-8ee6-964a36288fbc.png)

Without changing sky and bg color as well:
![clouds_fog_only](https://user-images.githubusercontent.com/2880643/27316988-ca07c92a-5583-11e7-97c9-ed7139389fc8.png)

bendeutsch commented:
Green clouds:
![clouds_green](https://user-images.githubusercontent.com/2880643/27317050-37183022-5584-11e7-9cf7-21e713c611e1.png)

Note the black stars, which are currently gated by the blue channel of the bg color. I'll probably have to change the gating, or disable stars when inside a cloud.

Fixer-007 commented:
Nice.

paramat commented:
Very nice, i assume not much different with shaders disabled.

bendeutsch commented:
@paramat Yes, it is almost exactly the same with shaders disabled. In fact, before I settled on changing the sky colors, I expected I would need to handle both paths separately, and started with the shaderless one because it seemed simpler. After using the sky colors and fog range, I guessed that this would work with shaders without any changes, so I enabled shaders and checked it out, and could not see a difference.

Shortly afterwards I started making screenshots, and only after I had snapped a few did I realize that these were with shaders on, not off…

bendeutsch commented:
The "bodies_visible" pull request above actually (temporarily) disables sun, moon and stars while inside a cloud. This may seem a bit much, but it did not feel wrong, so it might be a way to go. I also don't know much about how the sky works with an actual six-image sky box – does it disable stars too? If this conflicts with anything, I could change this back to a better way.

If not, I'd say it's feature complete, would clean up the code (thanks for the review @nerzhul !) and remove the WIP from the title.

bendeutsch commented:
Oops! I had just added some notes to the first comment / PR description, including
> Since it uses fog, the effect is completely negated when fog is turned off.
Turns out this is no longer the case! When fog is disabled, the sky still turns the color of clouds, and cloud rendering is disabled within a cloud. This seems wrong – what should I do when fog is disabled? Change the sky but still render the other clouds? Or just fall back to how it was without this PR?

paramat commented:
> fall back to how it was without this PR.

I think that ^ if fog is disabled. Changing sky colour when inside a coud with no fog effect will be weird.

bendeutsch commented:
That last commit caused the "inside a cloud" effect to depend on the current fog settings: if fog is disabled (F3 key by default for a temporary disable), then nothing happens. The decision whether to render the clouds if inside one has also been moved from `Clouds::render` to `Game::updateFrame`, since that is where the fog info resides.

I had almost had a commit with a `Cloud` member variable for "is fog enabled", but I think this solution separates the data better.

bendeutsch commented:
Rebased and squashed. Please have a look now.

paramat commented:
No longer WIP?

bendeutsch commented:
No longer WIP (I can't change labels – can I?).

I consider this finished. Please check it out, and if you like it, I'm hereby, well, requesting a pull :slightly_smiling_face: 

Change requests are welcome, of course, and will be included in the squashed commit.

rubenwardy commented:
I like this!

nerzhul commented on src/game.cpp:
```diff
@@ -4104,12 +4104,26 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
+						clouds->getColor());
+				sky->setBodiesVisible(false);
+				runData.fog_range = 20.0f * BS;
+				clouds->setVisible(
```
indent is very... unreadable here, can you improve it to inline some functions properlys?

bendeutsch replied:
Yes, I can, but this was generated by clang_format, so I'd probably be breaking some other formatting rule here.

bendeutsch replied:
Done, using a temporary variable. Please have a look.

paramat replied:
@nerzhul does this look ok to you now?

Fixer-007 commented:
I had problems with previous version of this PR: 1) when in cloud - weird thing on East/West 2) no clouds (even after rejoin). Will recheck.

bendeutsch commented:
@Fixer-007 :

1. I didn't touch the rendering of the "dawn glow" thing – you think I should suppress it as well when in a cloud? Actually, that makes sense…
2. Hm, ok, I'll need to fly around a bit. Shaders should not play a role for "inside cloud?" decisions.
3. The actual clouds should not be affected by this – this happens with this PR, but not with master?

Fixer-007 commented:
>Hm, ok, I'll need to fly around a bit. Shaders should not play a role for "inside cloud?" decisions.
Try to relogin many times when in cloud

>The actual clouds should not be affected by this – this happens with this PR, but not with master?
Only with this PR, there was another PR i tested as well, carpathian mapgen, I have no idea if it touches clouds in some ways.

paramat commented:
>  "dawn glow" thing – you think I should suppress it as well when in a cloud?

Yes please, it looks really bad from the imgur screenshots.

What happens with custom sun and moon textures in a texture pack?

nerzhul commented:
please rebase and finish the PR, we really want to see this merged this week :)

bendeutsch commented:
If I disable the glowy sky thing, I'll have disabled just about everything in the sky. I'll try to remove my "bodies visible" setting and just mark the sky as "not visible" when in a cloud. This should look "not fancy", but definitely "not broken", which is probably more important. On the other hand, this could cause undefined behaviour. I'll test a few things.

What happens to sun and moon textures depends on how they're implemented. If they're part of rendering the sky (specifically the sun and moon within `Sky::render`), they should disappear. This seems to be the case in `src/sky.cpp` at the moment.

paramat commented:
> just mark the sky as "not visible" when in a cloud

Sounds good to me.

bendeutsch commented:
I tried it with `sky->setVisible(false)`. This has several disadvantages:

1. You lose the "artificial horizon" caused by the "cloudy fog thing" near the horizon. This helps you keep your bearings.
2. The fog color is the background color (thanks to `bg_fallback_color`, I think), so you also lose the "all rendered things are slightly darker than the sky" effect, which I rather like.
3. The sky did not change back when I left a cloud! I'd have to retrace things to see what I need to do to fix this, no clue how these things interact.

No, I think I should leave the sky rendering in, just disable the drawing of "bodies", as before. Maybe I can even reorder code so that this becomes a fast return and I don't need half a dozen "ifs".

Then all that's left is to investigate the bug reports from above.

bendeutsch commented:
I've added a "fast abort" in `Sky::render` before sun, moon, stars and the orange glowy thing. Rebased, too.

@Fixer-007 I haven't been able to reproduce your points 2 and 3; I'll try some more, but if you could have another look, I'd be grateful.

Fixer-007 commented:
I will.

paramat commented:
> No, I think I should leave the sky rendering in, just disable the drawing of "bodies", as before

I agree.

Maybe bugs are caused by our 'camera offset' code that shifts the world every 200 nodes? #1119

bendeutsch commented:
@Fixer-007 Thanks, I also think I saw something similar during a test flight yesterday – it appeared far from the origin.

I'm relatively sure I did not change anything substantial about cloud rendering, but I'll take a closer look!

bendeutsch commented:
I can reproduce this now; current master is fine, this PR seems to lose clouds about 400 nodes out from  the origin :-( I had been pretty sure I had correctly taken the camera offset into account, seems not so much. I'll need to fix this.

bendeutsch commented:
Well, I goofed up :-( While passing the camera position as a `v3f`, and taking the camera offset of 3rd person views into account, I somehow dropped the camera offset thing.

I'm trying to get the rendering back to where it was, but for some reason, it still stops after 1800 nodes out or so. Very vexing. When I get that working again, it should be somewhat straightforward to recreate the "is camera inside cloud" detection in `Clouds::update`, since that *should* be a linear function of camera position and cloud origin – I hope…

bendeutsch commented:
There, that should fix it. Please have another look. I flew out to large positive and large negative X and Z coordinates, and it all worked.

Fixer-007 commented:
Will look.

bendeutsch commented:
Ah yes: the clouds (and the cloud effect) are only visible if the sky is visible – and there are circumstances in which it is "not visible", I take it. Can you fly just below the clouds, look up and down, and see if you can make them disappear then too? I'll see if I can find/make a world with a pit like that on my system here.

Fixer-007 commented:
Just use v7 mapgen, seed and position provided in gif and lower clouds.

paramat commented:
Fixer-007 that's a separate bug, clouds disappear whenever 'darkness detection' decides that sky is not visible. If you could see the sky it would be going dark at these times. So not an issue with this PR.

Fixer-007 commented:
Can confirm cloud is blinking even without this PR, that means it is separate issue not related to your code.

bendeutsch commented:
This PR has been retagged "WIP", but I no longer consider it such and don't see any outstanding problems, either.

Could someone please remove the WIP label? And is anything still stopping a merge? @nerzhul , you had some code changes; I think I fixed all of the first batch, but have lost your last change request (squashing on my side) – could you have another look?

paramat commented:
Tested and seems fine, day or night, with normal, coloured or glowing clouds.
One minor issue is that the upper half of the skybox becomes brighter when inside a cloud, although it is darker when outside, doesn't bother me much but just thought i'd mention it. Happens with any cloud colour.

![screenshot_20170705_121837](https://user-images.githubusercontent.com/3686677/27862323-8cbcb72e-617c-11e7-8eb5-aa4dbd489219.png)

paramat commented:
However the bright upper sky makes terrain visible which is a nice effect.

paramat commented:
Related to that, distant terrain is visible as a strong outline however distant it is, it doesn't fade into fog but disappears in blocks:

![screenshot_20170705_124402](https://user-images.githubusercontent.com/3686677/27863064-4eea798c-6180-11e7-95a0-c489f8daf983.png)

Again i don't see this as a problem, having distant outlines of terrain may even be a nice feature, but thought i'd mention it.

paramat commented on src/game.cpp:
```diff
@@ -4103,12 +4103,29 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
+				sky->setBodiesVisible(false);
+				runData.fog_range = 20.0f * BS;
+				clouds->setVisible(
+						false); // and don't draw clouds after all
```
Why split the line?

paramat commented:
:+1: Once that line is unsplit (if there isn't a reason for that).

paramat commented:
Please unsplit that line on merge.

SmallJoker commented:
bedb6d
